[RA] Disable split around hint register if optimize for size

weiguozhi opened the pull request:
Split a virtual register with hint may generate COPY instructions in multiple cold basic blocks, and increase code size. So disable this split when the function is optimized for size.

weiguozhi committed:
[RA] Disable split around hint register if optimize for size

Split a virtual register with hint may generate COPY instructions in
multiple cold basic blocks, and increase code size. So disable this
split when the function is optimized for size.

zmodem commented on llvm/lib/CodeGen/RegAllocGreedy.cpp:
```diff
@@ -1217,6 +1217,9 @@ bool RAGreedy::trySplitAroundHintReg(MCPhysReg Hint,
                                      const LiveInterval &VirtReg,
                                      SmallVectorImpl<Register> &NewVRegs,
                                      AllocationOrder &Order) {
+  if (MF->getFunction().hasOptSize())
```
We should probably add a comment explaining why (may increase code size).

zmodem commented on llvm/test/CodeGen/X86/no-split-size.ll:
```diff
@@ -0,0 +1,92 @@
+; @foo is optimized for size. Variables %p2, %p3, %p4, %p5 and %p6 are not split
+; in cold blocks.
+
+define i64 @foo(ptr %ptr, i64 %p2, i64 %p3, i64 %p4, i64 %p5, i64 %p6) #0 {
```
(I find just putting `optsize` directly instead of `#0` is easier to read.)

zmodem reviewed (approved):
Thanks! lgtm with nits.

weiguozhi committed:
[RA] Disable split around hint register if optimize for size

Split a virtual register with hint may generate COPY instructions in
multiple cold basic blocks, and increase code size. So disable this
split when the function is optimized for size.
